Conversations

jagerman opened the pull request:
Adds Conversation config messages.

Still TODO:
- [x] finish the test code
- [x] add actual implementations of the C API functions
- [x] prune >30-day old entries

jagerman committed:
Move session_id_is_valid into generic util header

jagerman committed:
Conversations config messages

WIP - the API is defined, but the test code is incomplete and the C API
does not yet have the implementations bridging the C API into C++.

jagerman committed:
Fix non-explicit contact_info constructor

mpretty-cyro commented on include/session/config/conversations.h:
```diff
@@ -0,0 +1,190 @@
+
+typedef struct convo_open_group {
+    const char* base_url; // null-terminated, always lower-case
+    const char* room; // null-terminated, always lower-case
```
Is the `room` always lower case? (currently iOS forces the `base_url` to lowercase but not the `room`)

mpretty-cyro replied:
Ah I saw the comment in `conversations.hpp` - feel free to just resolve this (I guess we will need to update the clients to force the `room` to lowercase? As whatever devices an open group wasn't joined on would have the value in lowercase regardless of what the casing was when initially joined)

jagerman replied:
I don't think this is our definitive list of which open groups you are in; that will likely live in a "groups" config message (and would be where the appropriate case is stored).  But it is case-insensitive on the SOGS side, so it seemed better to lower-case it here on input so that regardless of how a local client manages it, we'll have a consistent value here.

mpretty-cyro replied:
I guess I figured adding a separate "groups" config would be redundant since we have this (I'm not sure what else we'd store in there 😕)

jagerman replied:
I could see us storing some metadata in there such as the time that you joined, maybe a local alias for the group, and so on.

mpretty-cyro replied:
Yea ok, I guess my worry is that any functionality we add for open group conversations we'd also want for non-open-group conversations as well and just end up with `CONVERSATIONS` and `CONVERSATIONS_2` config types

jagerman replied:
My idea is that we have this one for tracking highly volatile conversation details (like last read timestamp), which we keep as small as possible, then put the details that don't change often into a "groups" config that contains the open group, closed group, legacy closed group fixed data.  (This is similar to what we have now with Contacts holding contact slow-changing data, and the last read timestamp going in here).

mpretty-cyro replied:
Mmk, did you want to shift the disappearing messages setting into that type instead of having it in here then? (It might also make sense to call that one `CONVERSATIONS` and rename this one to something like `CONVOS_LAST_READ` then)

mpretty-cyro replied:
(or are you thinking that the disappearing messages setting might change frequently enough as well?)

jagerman replied:
Although thinking about it some more, perhaps rather than storing URL + room + pubkey we should just store a (smaller) hash of that in here to keep size reduced.

So you'd go pull your list of open groups from the Groups config, then would look up H(url | room | pubkey) in here.

jagerman replied:
> Mmk, did you want to shift the disappearing messages setting into that type instead of having it in here then?

Yeah, actually, that's probably worthwhile.  I do want to keep this one (which gets re-push frequently) as small as possible.

mpretty-cyro replied:
No worries, I'll let the guys know that the conversation stuff is still in flux in case they happen to start working on it

jagerman committed:
Fix bugs; more c++ API tests; add single-type iterators

The single type iterators let you iterator over, for example, only
1-to-1 conversations.

jagerman committed:
Remove debugging headers

jagerman committed:
Finish C API, some tweaks to address minor issues

- open group URLs now have a length restriction of 320 characters for
  the base URL and 150 for the room name.  (These are needed so that we
  can use fixed buffers in the C API rather than make the C API require
  extra freeing of the returned open group data).
- abstracted some common code into src/config/internals.hpp
- finished C API implementations and added a C API version of the
  Conversations test code.

majestrate commented on include/session/config/conversations.h:
```diff
@@ -0,0 +1,222 @@
+///     convos_iterator_free(it);
+///
+
+typedef struct convos_iterator {
```
this iterator never is used as a non pointer so it can (and should) be a forward declared struct.

jagerman replied:
Why?  Just to make it opaque?

majestrate commented:
On Fri, 20 Jan 2023 14:58:39 -0800
Jason Rhinelander ***@***.***> wrote:

> @jagerman commented on this pull request.
> 
> 
> 
> > +///
> +///     convo_one_to_one c1;
> +///     convos_iterator *it = convos_iterator_new(my_convos);
> +///     while (!convos_iterator_done(it)) {
> +///         if (convo_it_is_1to1(it, &c1)) {
> +///             bool should_delete = /* ... */;
> +///             if (should_delete)
> +///                 convos_iterator_erase(it);
> +///             else
> +///                 convos_iterator_advance(it);
> +///         }
> +///     }
> +///     convos_iterator_free(it);
> +///
> +
> +typedef struct convos_iterator {
> 
> Why?  Just to make it opaque?
> 

yes, with a C api a sign of a good design is that things that CAN be
opaque are always made opaque.

the only time i'd see a need to expose internals like that is when we
need to know the size of the struct or we have public members on the
struct we want callers to get at.

none of those are in play on this one so it's better to make it opaque.

-- 
~jeff

majestrate commented:
On Sun, 22 Jan 2023 13:58:41 -0800
Jason Rhinelander ***@***.***> wrote:

> @jagerman commented on this pull request.
> 
> 
> 
> > +
> +
> +typedef struct convo_one_to_one {
> +    char session_id[67];  // in hex; 66 hex chars + null terminator.
> +
> +    // milliseconds since unix epoch:
> +    int64_t last_read;
> +
> +    // expiration mode & time:
> +    CONVO_EXPIRATION_MODE exp_mode;
> +    int64_t exp_minutes;
> +} convo_one_to_one;
> +
> +typedef struct convo_open_group {
> +    const char* base_url; // null-terminated, always lower-case
> +    const char* room; // null-terminated, always lower-case
> 
> Although thinking about it some more, perhaps rather than storing URL
> + room + pubkey we should just store a (smaller) hash of that in here
> to keep size reduced.
> 
> So you'd go pull your list of open groups from the Groups config,
> then would look up H(url | room | pubkey) in here.
> 

this kind of optimizatgion has caused issues in the past (desktop
session). i think doing a (pubkey, room) tuple as the lookup key would
be better but at least please omit the url part of it.

-- 
~jeff

jagerman committed:
Always use submodule for oxenc

Given how libsession-util is currently used, we likely need to include
oxenc headers, so makes little sense to try to use the system-installed
one.

jagerman committed:
Drop expiry, add unread to Conversations; add open group URL parsing

jagerman committed:
Rename Conversations -> ConvoInfoVolatile

jagerman committed:
Formatting

jagerman committed:
Bump macos deployment target

Targetting macos 10.12 prevents us from using variant

jagerman committed:
Fix missing ifdef around C++ extern closing

jagerman committed:
Old convo pruning

This adds pruning of last_read info for old conversations.  It does two
things:

- when a conversation is set with a last_read value more than 30 days
  ago we decline to actually add the conversation into the config.
- when we push, we first remove any conversations that have a last_read
  value more than 45 days.
  - This dual threshold overlap is so that clients don't ping-pong
    adding/removing a conversation if the clocks aren't in sync and the
    conversation is close to the threshold.
- We leave the record in place (regardless of the last_read timestamp)
  if the manual `unread` flag is set for the conversation.

jagerman committed:
Test suite fix with pruning

This contact was getting pruned now (since it has neither last_read nor
unread) which broke the test.  Set unread on it to fix.

jagerman committed:
Fix dirty->merge->dirty mutability bug

If you have a dirty config and merged pushed changes (before pushing
yourself) you'd end up with a non-mutable internal config object that is
marked dirty.  It *should* be dirty (because the dirtyness carries
forward from before the merge), but it should also be Mutable because
dirty config objects are always expected/required to be mutable.

This fixes it by properly constructing a MutableConfigMessage instead of
just a ConfigMessage when we merge configs into an already-dirty config
so that we properly remain mutable.

jagerman committed:
Restructure convo volatile storage to not repeat SOGS server/pubkey

This rearranges how we store pubkey data so that, rather than storing
as:

    "o": {
        "https://some.long.server\0ROOM1\0abcdef0123456789...": { ... },
        "https://some.long.server\0ROOM2\0abcdef0123456789...": { ... },
        "https://some.long.server\0ROOM3\0abcdef0123456789...": { ... },
        "http://example.com:12345\0ROOM1\01234567890abcdef...": { ... },
    }

we now store as:

    "o": {
        "https://some.long.server": {
            "#": "abcdef0123456789...",
            "R": {
                "ROOM1": { ... },
                "ROOM2": { ... },
                "ROOM3": { ... }
            }
        },
        "http://example.com:12345": {
            "#": 01234567890abcdef...",
            "R": {
                "ROOM1": { ... }
            }
        }
    }

which avoids substantial duplication when a user is in multiple rooms on
the same server.

This also unifies with the storage layout of open groups in the upcoming
group list config type.

jagerman committed:
Formatting

jagerman committed:
Add libzstd to static bundle

frtget commented on include/session/config/convo_info_volatile.hpp:
```diff
@@ -0,0 +1,414 @@
+///
+/// o - open group conversations.  This is a nested dict where the outer keys are the BASE_URL of
+///     the open group and the outer value is a dict containing:
+///     - `#` -- the 32-byte server pubkey
```
Would it make sense to swap the BASE_URL and the pubkey to prevent duplicates since a sogs can be reached via multiple BASE_URLs ?

frtget replied:
I'm assuming changing the pubkey of a sogs would break things

majestrate replied:
>I'm assuming changing the pubkey of a sogs would break things

yes, many things would. the stealth addresses ( the `15` prefixed ones ) of all the users would be the most notable of those changes.

jagerman replied:
> Would it make sense to swap the BASE_URL and the pubkey to prevent duplicates since a sogs can be reached via multiple BASE_URLs ?

No: You can use the same pubkey on multiple SOGS instances—and you might want to if, for instance, you decided to split up a single SOGS to multiple SOGS.  But it is impossible to use different pubkeys at the same URL.

majestrate commented on include/session/config/convo_info_volatile.h:
```diff
@@ -0,0 +1,219 @@
+///             if (should_delete)
+///                 convo_info_volatile_iterator_erase(it);
+///             else
+///                 convo_info_volatile_iterator_advance(it);
```
this advancement of the iterator should be done outside the `if(convo_it_is_1to1(...` block as well.

jagerman replied:
Fixed in #15
